executor/src/witgen/jit/function_cache.rs

```diff
@@ -0,0 +1,139 @@
+    /// but failed.
+    witgen_functions: BTreeMap<CacheKey, Option<WitgenFunction<T>>>,
+    column_layout: ColumnLayout,
+}
+
+impl<'a, T: FieldElement> FunctionCache<'a, T> {
+    pub fn new(
+        fixed_data: &'a FixedData<'a, T>,
+        parts: MachineParts<'a, T>,
+        block_size: usize,
+        latch_row: usize,
+        metadata: ColumnLayout,
+    ) -> Self {
+        let processor =
+            BlockMachineProcessor::new(fixed_data, parts.clone(), block_size, latch_row);
+
+        FunctionCache {
+            processor,
+            column_layout: metadata,
+            witgen_functions: BTreeMap::new(),
+        }
+    }
+
+    /// Compiles the JIT function for the given identity and known arguments.
+    /// Returns true if the function was successfully compiled.
+    pub fn compile_cached(&mut self, identity_id: u64, known_args: &BitVec) -> bool {
+        let cache_key = CacheKey {
+            identity_id,
+            known_args: known_args.clone(),
+        };
+        self.ensure_cache(&cache_key);
+        self.witgen_functions.get(&cache_key).unwrap().is_some()
+    }
+
+    fn ensure_cache(&mut self, cache_key: &CacheKey) {
+        if self.witgen_functions.contains_key(cache_key) {
+            return;
+        }
+
+        let f = match T::known_field() {
+            // Currently, we only support the Goldilocks fields
+            Some(KnownField::GoldilocksField) => self.compile_witgen_function(cache_key),
+            _ => None,
+        };
+        assert!(self.witgen_functions.insert(cache_key.clone(), f).is_none())
+    }
+
+    fn compile_witgen_function(&self, cache_key: &CacheKey) -> Option<WitgenFunction<T>> {
+        log::trace!("Compiling JIT function for {:?}", cache_key);
+        self.processor
+            .generate_code(cache_key.identity_id, &cache_key.known_args)
+            .ok()
+            .and_then(|code| {
+                log::trace!("Generated code ({} steps)", code.len());
+                let known_inputs = cache_key
+                    .known_args
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, b)| if b { Some(Variable::Param(i)) } else { None })
+                    .collect::<Vec<_>>();
+
+                if code
+                    .iter()
+                    .any(|effect| matches!(effect, Effect::MachineCall(_, _)))
+                {
+                    // TODO: Machine calls trigger a unimplemented!() in compile_effects()
+                    return None;
+                }
+
+                log::trace!("Compiling effects...");
+
+                Some(
+                    compile_effects(
+                        self.column_layout.first_column_id,
+                        self.column_layout.column_count,
+                        &known_inputs,
+                        &code,
+                    )
+                    .unwrap(),
```

Comment:
Unwrapping here means that any compilation error will fail hard, including if the Rust compiler is not set up properly on the user's machine. Maybe that's too tough an I should instead return `compile_effects(...).ok()`. It is nice to see that the tests pass with `unwrap()` though.

Reply:
I think we should hard fail unless the cargo is not installed. I think we should be able to detect that from `Command`.

Reply:
Hmm, seems kinda complicated, mostly because I don't even know how to test that easily. I added a TODO comment, failed softly (so we don't break for users without `cargo`) and would prefer to leave this for a future PR.